cpp/src/arrow/compute/kernels/hash.cc

```diff
@@ -415,5 +506,27 @@ Status DictionaryEncode(FunctionContext* ctx, const Datum& value, Datum* out) {
   return Status::OK();
 }
 
+Status ValueCounts(FunctionContext* ctx, const Datum& value,
+                   std::shared_ptr<Array>* counts) {
+  std::unique_ptr<HashKernel> func;
+  RETURN_NOT_OK(GetValueCountsKernel(ctx, value.type(), &func));
+
+  // Calls return nothing for counts.
+  std::vector<Datum> unused_output;
+  std::shared_ptr<Array> uniques;
+  RETURN_NOT_OK(InvokeHash(ctx, func.get(), value, &unused_output, &uniques));
+
+  Datum value_counts;
+  RETURN_NOT_OK(func->FlushFinal(&value_counts));
+
+  auto data_type = std::make_shared<StructType>(std::vector<std::shared_ptr<Field>>{
+      std::make_shared<Field>("Values", uniques->type()),
```

Comment:
Make `Values` and `Count` a public `constexpr` such that user can programatically retrieve the field by name at compile time, maybe expose the index that way too.

Reply:
Good suggestion.  Done.